exonum/src/runtime/rust/stubs.rs

```diff
@@ -0,0 +1,236 @@
+    use super::*;
+    use exonum_crypto::gen_keypair;
+    use exonum_merkledb::BinaryValue;
+
+    // Suppose we have the following trait describing user service.
+    trait Token<Ctx> {
+        type Output;
+
+        fn create_wallet(&self, context: Ctx, wallet: CreateWallet) -> Self::Output;
+        fn transfer(&self, context: Ctx, transfer: Transfer) -> Self::Output;
+    }
+
+    // The `Ctx` type param allows to provide additional information to the implementing type.
+    // For example, many stubs require to know the instance ID to which the call is addressed.
+    // For these stubs `Ctx == InstanceId` may make sense. In other cases, the context
+    // may be void `()`.
+
+    // We don't quite care about types here, so we define them as:
+    type CreateWallet = String;
+    type Transfer = u64;
+    // In general, we accept any type implementing the `BinaryValue` trait.
+
+    // Our goal is to provide an implementation of this user-defined trait for some generic
+    // types, e.g., a keypair (which would generate signed transactions when called), or
+    // `CallContext` (which would call another service on the same blockchain).
+
+    // In order to accomplish this, we notice that for all possible service traits,
+    // there exists a uniform conversion of arguments: the argument (i.e.,
+    // `wallet` for `create_wallet`, `transfer` for `transfer`) can always be converted to
+    // a `Vec<u8>` since it implements the `BinaryValue` trait. Moreover, this conversion
+    // is performed by the stub types anyway (e.g., the keypair needs to get the binary serialization
+    // of the message in order to create a signature on it).
+
+    // Similarly, the information about the method itself is also uniform; it consists of
+    // the method ID and name. This info is encapsulated in the `MethodDescriptor` type
+    // in the parent module.
+
+    // The existence of uniform conversions gives us an approach to the solution. We need
+    // to define a more generic trait (`GenericCall` / `GenericCallMut`), which would then
+    // be implemented for any user-defined service interface like this:
+    impl<T, Ctx> Token<Ctx> for T
+    where
+        T: GenericCall<Ctx>,
+    {
+        type Output = <T as GenericCall<Ctx>>::Output;
+
+        fn create_wallet(&self, context: Ctx, wallet: CreateWallet) -> Self::Output {
+            const DESCRIPTOR: MethodDescriptor<'static> = MethodDescriptor {
+                interface_name: "",
+                name: "create_wallet",
+                id: 0,
+            };
+            self.generic_call(context, DESCRIPTOR, wallet.into_bytes())
+        }
+
+        fn transfer(&self, context: Ctx, transfer: Transfer) -> Self::Output {
+            const DESCRIPTOR: MethodDescriptor<'static> = MethodDescriptor {
+                interface_name: "",
+                name: "transfer",
+                id: 1,
+            };
+            self.generic_call(context, DESCRIPTOR, transfer.into_bytes())
```

Comment:
Does it make sense to create a macro to reduce the repeated code ?

Reply:
Do you mean in the `explanation` module? In real interfaces, the similar code [is generated](https://github.com/exonum/exonum/pull/1606/files#diff-4f9854381ab5abbdcabc17fb9bafa3a2R296) by the `exonum_interface` proc macro.